Fix text selection handles showing outside the visible text region

sjindel-google opened the pull request:
This follows the Android behavior that the text selection handles fade in and out when they enter/leave the viewport of the text, remaining clamped within the viewport and not drawing outside of it.

This also fixes an issue with gestures being incorrectly shifted after scrolling.

fixes #24214

sjindel-google committed:
Don't show handles outside the text field's boundary.

Also fix 27948.

justinmc reviewed (dismissed):
Just a few small comments here, but someone else more knowledgeable should take a look at the overall approach to `PaintVisibility` etc.

I wanted to point out some slight difference in behavior from native Android as well.  This PR continues to show the context menu even when both handles are hidden, but natively the context menu disappears in that case.

<img width="307" alt="screen shot 2019-01-04 at 4 53 24 pm" src="https://user-images.githubusercontent.com/389558/50680125-78603380-1041-11e9-8227-ca953ed45dc0.png">

Also, natively there is a fade animation when a handle disappears/reappears that we might want to consider adding to this PR.

![handles](https://user-images.githubusercontent.com/389558/50680613-4780fe00-1043-11e9-93fa-b07ad6d666dc.gif)

I mentioned this in the related issue, ~~but iOS also might be a concern for this bug.  Natively it's not possible to cause a selection to not be visible, but in our implementation in iOS it looks like it is.~~ iOS is a concern for this bug as well.  It's possible to scroll the handles off screen in a native iOS app.

I'm not sure how accurate we want to be in this PR or if we want to put off minor details like those to future PRs.

GaryQian commented:
cc @Hixie for potential API additions

tvolkert commented:
Also +cc @goderbauer

goderbauer reviewed (commented):
Thinking out loud: Why can we not just use a regular setState to change the visibility of a `Visibility` widget to hide the selectors when the textfield determines that they will not be visible anymore? Sure, we may be a frame late, but since the fade-out is supposed to be animated anyways it may not be noticeable?

goderbauer commented:
@sjindel-google I know this PR is pretty old and I apologize for the delay in reviewing it. Is this still something you plan to work on? Or should we close it for now? I would like to remove stale PRs from our backlog so we can get to new ones quicker in the future.

sjindel-google commented:
> @sjindel-google I know this PR is pretty old and I apologize for the delay in reviewing it. Is this still something you plan to work on? Or should we close it for now? I would like to remove stale PRs from our backlog so we can get to new ones quicker in the future.

I will work on it. Sorry that I haven't been able to respond to your comments yet, but I'm back from vacation/offsite on Monday and will pick it up around then.

sjindel-google commented:
> Thinking out loud: Why can we not just use a regular setState to change the visibility of a `Visibility` widget to hide the selectors when the textfield determines that they will not be visible anymore? Sure, we may be a frame late, but since the fade-out is supposed to be animated anyways it may not be noticeable?

I tried this approach first, but I hit an assert in `_handleBuildScheduled()`:

```dart
    assert(() {
      if (debugBuildingDirtyElements) {
        throw FlutterError(
          'Build scheduled during frame.\n'
          'While the widget tree was being built, laid out, and painted, '
          'a new frame was scheduled to rebuild the widget tree. '
          'This might be because setState() was called from a layout or '
          'paint callback. '
          'If a change is needed to the widget tree, it should be applied '
          'as the tree is being built. Scheduling a change for the subsequent '
          'frame instead results in an interface that lags behind by one frame. '
          'If this was done to make your build dependent on a size measured at '
          'layout time, consider using a LayoutBuilder, CustomSingleChildLayout, '
          'or CustomMultiChildLayout. If, on the other hand, the one frame delay '
          'is the desired effect, for example because this is an '
          'animation, consider scheduling the frame in a post-frame callback '
          'using SchedulerBinding.addPostFrameCallback or '
          'using an AnimationController to trigger the animation.'
        );
      }
      return true;
    }());
```

goderbauer commented:
Looking at @justinmc's earlier comment, isn't the AnimationController mentioned in that assert exactly what you want to get around that limitation to fade the controls in/out over the next couple of frames?

sjindel-google commented:
> Looking at @justinmc's earlier comment, isn't the AnimationController mentioned in that assert exactly what you want to get around that limitation to fade the controls in/out over the next couple of frames?

I'll give that approach a shot.

sjindel-google commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -60,6 +60,11 @@ enum SelectionChangedCause {
 /// Used by [RenderEditable.onCaretChanged].
 typedef CaretChangedHandler = void Function(Rect caretRect);
 
+/// Signature for the callback that reports when the visibility of either extent
```
Done.

sjindel-google commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -227,6 +233,41 @@ class RenderEditable extends RenderBox {
+  bool _startSelectionExtentVisible = true;
+  bool _endSelectionExtentVisible = true;
+
+  /// Invoked whenever the visbility of the start or end of the selection
```
Done.

sjindel-google commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5835,3 +5835,59 @@ class _StatefulBuilderState extends State<StatefulBuilder> {
+  /// paint phase. The [visibility] argument must not be null.
+  const PaintVisibility({
+    Key key,
+    @required
```
I've removed `PaintVisibility` and `RenderVisibility`.

sjindel-google commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -60,6 +60,11 @@ enum SelectionChangedCause {
 /// Used by [RenderEditable.onCaretChanged].
 typedef CaretChangedHandler = void Function(Rect caretRect);
 
+/// Signature for the callback that reports when the visibility of either extent
```
Done.

sjindel-google commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5835,3 +5835,59 @@ class _StatefulBuilderState extends State<StatefulBuilder> {
+/// the object is scrolled out of view.
+class PaintVisibility extends SingleChildRenderObjectWidget {
+  /// Creates a widget whose child's visibility can be controlled during the
+  /// paint phase. The [visibility] argument must not be null.
```
I've removed `PaintVisibility`.

sjindel-google commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5835,3 +5835,59 @@ class _StatefulBuilderState extends State<StatefulBuilder> {
+
+  ValueListenable<bool> visibility = ValueNotifier<bool>(true);
+
+  @override void paint(PaintingContext context, Offset offset) {
```
I've removed `PaintVisibility`.

sjindel-google commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -221,6 +222,9 @@ class TextSelectionOverlay {
   ///
   /// The [context] must not be null and must have an [Overlay] as an ancestor.
+  ///
+  /// 'selectionStartInView' and 'selectionEndInView' are optional and can be
```
I noticed that the `_TextSelectionHandleOverlayState` has direct access to the `RenderEditable`, so I've removed these callbacks and typedefs since it's not needed to thread this through `EditableTextState` and `TextSelectionOverlay`.

Hixie replied:
Looks like the documentation wasn't removed at the same time

sjindel-google commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5835,3 +5835,59 @@ class _StatefulBuilderState extends State<StatefulBuilder> {
+  }) : super(key: key, child: child);
+
+  /// Controls whether the child is visible.
+  final ValueListenable<bool> visibility;
```
Fixed, now the AnimationController will be triggered by it (in another file).

sjindel-google commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5835,3 +5835,59 @@ class _StatefulBuilderState extends State<StatefulBuilder> {
+///
+/// This allows the visibility of a child to be determined at paint time, which
+/// cannot be achieved with [Visibility] because the paint phase happens after
+/// the build and layout phases. Make sure that this widget will be repained
```
I've removed `PaintVisibility` so this is not a concern anymore.

sjindel-google commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5835,3 +5835,59 @@ class _StatefulBuilderState extends State<StatefulBuilder> {
+///
+/// This is used for example to hide an overlay which points to an object when
+/// the object is scrolled out of view.
+class PaintVisibility extends SingleChildRenderObjectWidget {
```
I've removed `PaintVisibility` so this is not a concern anymore.

sjindel-google commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5835,3 +5835,59 @@ class _StatefulBuilderState extends State<StatefulBuilder> {
+  }
+}
+
+class _RenderVisibility extends RenderProxyBox {
```
I've removed `PaintVisibility` so this is not a concern anymore.

sjindel-google commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -5835,3 +5835,59 @@ class _StatefulBuilderState extends State<StatefulBuilder> {
+/// This is used for example to hide an overlay which points to an object when
+/// the object is scrolled out of view.
+class PaintVisibility extends SingleChildRenderObjectWidget {
+  /// Creates a widget whose child's visibility can be controlled during the
```
I've removed `PaintVisibility` so this is not a concern anymore.

sjindel-google commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -1979,6 +1979,81 @@ testWidgets(
     ));
     expect(called, 2);
   });
+
```
I've removed `PaintVisibility`.

sjindel-google commented:
I've removed `PaintVisibility` in favor of the `AnimationController` solution, and implemented animation.
I've also merged it with the animation for fading in the handles when they first appear.

goderbauer commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -10,6 +10,7 @@ import 'package:flutter/services.dart';
 import 'package:flutter/scheduler.dart';
 import 'package:flutter/gestures.dart' show DragStartBehavior;
 
+import '../foundation/change_notifier.dart';
```
import should be :`package:flutter/foundation.dart`

sjindel-google replied:
Done.

goderbauer commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -372,33 +372,29 @@ class TextSelectionOverlay {
+      selectionControls: selectionControls,
+      position: position,
+      dragStartBehavior: dragStartBehavior,
+      vsync: Overlay.of(context),
```
Can you not just get this with `Overlay.of` from within the _TextSelectionHandleOverlay?

sjindel-google replied:
Removed since using `SingleTickerProviderStateMixin`.

goderbauer commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -546,31 +585,36 @@ class _TextSelectionHandleOverlayState extends State<_TextSelectionHandleOverlay
         break;
     }
 
+    point = widget.renderObject.clampToViewport(point);
```
For my own understanding: Why is this necessary?

sjindel-google replied:
This prevents the handles from drawing outside the box while they're fading out:

![before](https://user-images.githubusercontent.com/29863079/52866283-a2c40700-313e-11e9-916d-abcbc9075b95.gif)

Instead we get:

![after](https://user-images.githubusercontent.com/29863079/52866290-a5bef780-313e-11e9-8e5d-3c5c0041ce48.gif)

The second is more similar to the native behavior on Android, and I think it looks better than the first.

Hixie commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -249,6 +249,41 @@ class RenderEditable extends RenderBox {
+  /// "World", then scrolls so only "Hello" is visible, this will become
+  /// 'false'. If the user scrolls back so that the "d" is visible again, this
+  /// will become 'true'.
+  final ValueNotifier<bool> selectionEndInView = ValueNotifier<bool>(true);
```
we should probably add a paragraph to these two defining "visibility" as meaning "scrolled so that the selection is within the text field viewport", as opposed to being literally visible on the screen.

sjindel-google replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -249,6 +249,41 @@ class RenderEditable extends RenderBox {
+    final bool endExtentVisible = visibleRegion.contains(
+        _textPainter.getOffsetForCaret(
+            TextPosition(offset: _selection.end, affinity: _selection.affinity),
+            Rect.zero) +
```
nit: flutter style; see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#if-you-have-a-newline-after-some-opening-punctuation-match-it-on-the-closing-punctuation

sjindel-google replied:
Done.

Hixie commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -487,6 +485,47 @@ class _TextSelectionHandleOverlay extends StatefulWidget {
+  void initState() {
+    super.initState();
+
+    _controller = AnimationController(duration: _fadeDuration, vsync: widget.vsync)..value = 0.0;
```
value should start at 0.0, no need to set it

sjindel-google replied:
Done.

Hixie commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -468,6 +464,7 @@ class _TextSelectionHandleOverlay extends StatefulWidget {
     @required this.onSelectionHandleChanged,
     @required this.onSelectionHandleTapped,
     @required this.selectionControls,
+    @required this.vsync,
```
rather than taking a vsync, use TickerProviderSingleStateMixin

sjindel-google replied:
Done. For my own enlightenment, why is this better? I was previously just using the same ticker provider as the existing fade-in animation.

Hixie commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -1998,6 +1998,65 @@ void main() {
+    verifyVisibility(false, false);
+
+    // TODO(sjindel): Test dragging in the other direction (blocked by
+    // https://github.com/flutter/flutter/issues/27948).
```
please embrace the yak shave and fix that bug too, so that you can write the test

sjindel-google replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -1029,6 +1064,26 @@ class RenderEditable extends RenderBox {
+        break;
+    }
+    return point;
+  }
```
you should test this logic works.
for example, have a test that:
1. selects a word, verifies that opacity is 0.0, waits a second, verifies that opacity is 1.0.
2. scrolls the text field, keeping the handle in the text field, verifies that opacity is 1.0 and in the right place.
3. scrolls the text field, move the handle out. verify that opacity is 1.0 and in the right place (not outside field)
4. waits a tiny bit, verifies that opacity is 0.5 and still in the same place
5. scroll the handle back in, check that opacity is 0.5 but it's no longer stuck on the edge
6. etc

sjindel-google replied:
I've updated the current test to do check the positioning and animations.

Hixie commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -249,6 +249,41 @@ class RenderEditable extends RenderBox {
+  /// "Hello", then scrolls so only "World" is visible, this will become
+  /// 'false'. If the user scrolls back so that the "H" is visible again, this
+  /// will become 'true'.
+  final ValueNotifier<bool> selectionStartInView = ValueNotifier<bool>(true);
```
you should expose a `ValueListenable<bool>` and have the `ValueNotifier` be private. Otherwise other code can affect the value.

sjindel-google replied:
Done.

goderbauer reviewed (commented):
I like this better :)

sjindel-google commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -249,6 +249,41 @@ class RenderEditable extends RenderBox {
+  /// "Hello", then scrolls so only "World" is visible, this will become
+  /// 'false'. If the user scrolls back so that the "H" is visible again, this
+  /// will become 'true'.
+  final ValueNotifier<bool> selectionStartInView = ValueNotifier<bool>(true);
```
Given Hixie's remark above, and the new explanation in the comment, I think this name more precise. I've changed "InView" to "InViewport" to be even more clear.

sjindel-google commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -249,6 +249,41 @@ class RenderEditable extends RenderBox {
+    final Rect visibleRegion = Offset.zero & size;
+
+    final bool startExtentVisible = visibleRegion.contains(
+        _textPainter.getOffsetForCaret(
```
Done.

sjindel-google commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -1029,6 +1064,26 @@ class RenderEditable extends RenderBox {
     return null;
   }
 
+  /// Clamps a local coordinate within the viewport.
```
Done.

sjindel-google commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -1029,6 +1064,26 @@ class RenderEditable extends RenderBox {
+      case Axis.horizontal:
+        final double dx = point.dx.clamp(0.0, _viewportExtent);
+        if (dx != point.dx)
+          return Offset(point.dx.clamp(0.0, _viewportExtent), point.dy);
```
Done.

sjindel-google commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -1029,6 +1064,26 @@ class RenderEditable extends RenderBox {
+      case Axis.horizontal:
+        final double dx = point.dx.clamp(0.0, _viewportExtent);
+        if (dx != point.dx)
+          return Offset(point.dx.clamp(0.0, _viewportExtent), point.dy);
```
As far as I understand, this isn't possible currently if the input coordinate is coming from `RenderEditable.getEndpointsForSelection`, because the text only scrolls in one direction (either vertically or horizontally, but not both), and therefore a handle can only be outside the viewport along one axis.

However, I realized that the code is simpler is I just use `size.contains` from the enclosing `RenderEditable`, so I've just removed this function altogether.

sjindel-google commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -222,6 +223,9 @@ class TextSelectionOverlay {
   ///
   /// The [context] must not be null and must have an [Overlay] as an ancestor.
+  ///
+  /// 'selectionStartInView' and 'selectionEndInView' are optional and can be
```
Removed.

Hixie commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -1191,7 +1235,7 @@ class RenderEditable extends RenderBox {
   /// down events by calling this method.
   void handleTapDown(TapDownDetails details) {
-    _lastTapDownPosition = details.globalPosition - _paintOffset;
+    _lastTapDownPosition = details.globalPosition;
```
I'm surprised that this change doesn't have a corresponding number of other changes doing the opposite. What's the story here?

sjindel-google replied:
I thought it was not tested because the selection when scrolling was completely broken, but when I dug further I discovered that it was broken in commit 892c8919c7, where the correct behavior was implemented in `selectWordEdge()`, which is used on iOS, but broken for Android, which uses `selectPositionAt()` and `selectWordsInRange()`.

The `_paintOffset` is already subtracted in `selectPositionAt()` and `selectWordsInRange()`, but was not subtracted in `selectWordEdge()`. So I've updated `selectWordEdge()`. All the uses now correct subtract `_paintOffset` exactly once.

goderbauer replied:
Can you add some specific test cases that cover this (outside of their use in the selection handle test cases) to ensure that we never break this again?

sjindel-google replied:
Working on this. I wrote a test which uncovered a strange issue with taps not being received when they should be.

sjindel-google replied:
It looks the original issue I was solving here (#27948) was already fixed in (#28242).
The "issue" I ran into in the test was just a "user-error" caused by the xster's refactoring of the gesture logic.

Hixie commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -484,9 +478,50 @@ class _TextSelectionHandleOverlay extends StatefulWidget {
+    if (_visibility.value) {
+      _controller.forward();
+    }
+    _visibility.addListener(_handleVisibilityChanged);
```
you don't handle the case of `widget.position` changing.

You probably want a didUpdateWidget that unsubscribes from the old one and subscribes to the new one if it changed.

sjindel-google replied:
For my own enlightenment, how can this happen? When the selection handles are built, either none are built, only the left handle, or both handles. I don't see how it can happen that the left handle switches to use state from the right handle or vice versa.

goderbauer replied:
At the end of the day _TextSelectionHandleOverlay is just another regular widget and it should conform to the widget protocol, which allows that a state object can be attached to an arbitrary new configuration (e.i. Widget). This new Widget may have a different position.

sjindel-google replied:
Thanks for the explanation. Done.

Hixie commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -546,31 +581,42 @@ class _TextSelectionHandleOverlayState extends State<_TextSelectionHandleOverlay
+        point.dx.clamp(0.0, viewport.width),
+        point.dy.clamp(0.0, viewport.height),
+      );
+    }
```
Seems like it'd be cheaper to just always clamp?

sjindel-google replied:
I considered that, but I imagined that the "common case" is where both handles are visible, in which case the clamping doesn't change the input offset. It seemed better to me to avoid unnecessary allocation in this case.

goderbauer replied:
It would also be easier to read if we always clamp and allocations are supposed to be cheap in dart... If you have a benchmark showing that this is better, we can keep it. Otherwise I'd suggest changing it to the simplified version.

sjindel-google replied:
Done.

goderbauer reviewed (commented):
I think this looks good once the last two open comments are resolved.

sjindel-google committed:
Always clamp

sjindel-google commented:
It looks like #27948 was already fixed in #28242.

sjindel-google committed:
Update test

goderbauer commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -562,31 +603,41 @@ class _TextSelectionHandleOverlayState extends State<_TextSelectionHandleOverlay
     }
 
+    final Size viewport = widget.renderObject.size;
+    point = Offset(
```
nit: flutter style would be to add a trailing comma and indent by just two spaces:

```dart
point = Offset(
  point.dx.clamp(0.0, viewport.width),
  point.dy.clamp(0.0, viewport.height),
);
```

sjindel-google replied:
Done.

goderbauer reviewed (approved):
LGTM

Thanks for your patience and sticking with this PR

sjindel-google commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -255,6 +255,50 @@ class RenderEditable extends RenderBox {
+  ///
+  /// For example, if the text contains "Hello World", and the user selects
+  /// "Hello", then scrolls so only "World" is visible, this will become
+  /// 'false'. If the user scrolls back so that the "H" is visible again, this
```
Done.

sjindel-google committed:
Code style

sjindel-google commented:
> LGTM
> 
> Thanks for your patience and sticking with this PR

No, thank *you* and @Hixie for *your* patience!
